In [1]:
# SageMath version at least 10.4
%display latex
version()

'SageMath version 10.4, Release Date: 2024-07-19'

## Proof of Theorem 4.4 - Part 1

We have to prove that if $F = F_r$ is the homogeneous polynomial of degree $4$ computed in the notebook `Stiefel-manifold.ipynb` then 
$$
F_r(x_0, \ldots , x_{21}) \ge 0 \qquad \text{ for } \qquad 0 < r \le \frac49, \, x_0, \ldots, x_{21} \in \mathbb R
$$

Since $F_r(x_0, \ldots, x_{21})$ is linear on $r$, it is enough to prove that $F_0$ and $F_{\frac 49}$ are non-negative polynomials. We will do this by showing that $F_r$ is a SOS (sums of squares) polynomial for $r = 0, \frac 49$. By the results in Powers-Wörmann 1998 this is equivalent to the existence of a symmetric positive semi-definite matrix $H_r$ such that 
$$
F_r(x_0, \ldots, x_{21}) = \boldsymbol{x}^T \, H_r \, \boldsymbol{x}, \qquad r = 0, \frac49 \qquad \qquad (*)
$$
where $\boldsymbol x$ is the vector of monomials of degree $2$ in the variables $x_0, \ldots, x_{21}$. Since every monomial with non-trivial coefficient in $F_r$ has the form $x_i x_j x_k x_l$ with $0 \le i \le j \le 10 < k \le l \le 21$, we can replace $\boldsymbol x$ by the smaller vector of monomials
$$
\boldsymbol{x}^T = (x_0 x_{11}, \ldots, x_0 x_{21}, \ldots, x_{10} x_{11}, \ldots, x_{10} x_{21}, x_0^2, \ldots, x_{21}^2).
$$

In this notebook we compute all the symmetric (not necessarily positive semi-definite) matrices $H_r$ which satisfy $(*)$.

In [2]:
n = 22
r = var("r")
x_vars = [var("x"+str(i)) for i in range(n)]
F = load("F.sobj")
print(F)

1/2*(9*r - 4)*x0*x1*x11*x12 - 1/4*((9*r - 4)*x1^2 - 2*x10^2)*x11^2 - 1/4*((9*r - 4)*x0^2 - 2*x10^2)*x12^2 + 3/4*(r*x0^2 + r*x1^2 + r*x10^2)*x13^2 + 1/2*(x0^2 + x1^2 + x10^2)*x14^2 + 1/2*(x0^2 + x1^2)*x15^2 + 1/2*(x0^2 + x1^2)*x16^2 + 1/2*(x0^2 + x1^2 + x10^2)*x17^2 - 1/4*((9*r - 8)*x10^2 - 2*x0^2 - 2*x1^2)*x18^2 + 1/2*(x0^2 + x1^2 + x10^2)*x19^2 - 3/2*(r*x0*x11 + r*x1*x12)*x13*x2 + 3/4*(r*x11^2 + r*x12^2 + r*x14^2 + r*x15^2 + r*x16^2 + r*x17^2 + r*x18^2 + r*x19^2)*x2^2 + 1/4*(3*r*x2^2 + 2*x0^2 + 2*x1^2 + 2*x10^2)*x20^2 + 1/4*(3*r*x2^2 + 2*x0^2 + 2*x1^2)*x21^2 + 1/4*(3*r*x13^2 - (9*r - 8)*x17^2 + 2*x11^2 + 2*x12^2 + 2*x15^2 + 2*x16^2 + 2*x18^2 + 2*x21^2)*x3^2 + 1/4*(3*r*x13^2 - (9*r - 8)*x16^2 + 2*x11^2 + 2*x12^2 + 2*x14^2 + 2*x17^2 + 2*x19^2 + 2*x20^2)*x4^2 + 1/4*(3*r*x13^2 - (9*r - 8)*x15^2 + 2*x11^2 + 2*x12^2 + 2*x14^2 + 2*x17^2 + 2*x19^2 + 2*x20^2)*x5^2 + 1/4*(3*r*x13^2 - (9*r - 8)*x14^2 + 2*x11^2 + 2*x12^2 + 2*x15^2 + 2*x16^2 + 2*x18^2 + 2*x21^2)*x6^2 + 1/4*(3*r*x13^2 - (9*r - 8)*x

In [3]:
# Monomials of degree 2
X_mon2 = vector([x_vars[i]*x_vars[j] for i in range(n/2) for j in range(n/2,n)]+[x_vars[i]^2 for i in range(n)])
X_mon2_len = len(X_mon2)
print(X_mon2_len)
print("")
print(X_mon2)

143

(x0*x11, x0*x12, x0*x13, x0*x14, x0*x15, x0*x16, x0*x17, x0*x18, x0*x19, x0*x20, x0*x21, x1*x11, x1*x12, x1*x13, x1*x14, x1*x15, x1*x16, x1*x17, x1*x18, x1*x19, x1*x20, x1*x21, x11*x2, x12*x2, x13*x2, x14*x2, x15*x2, x16*x2, x17*x2, x18*x2, x19*x2, x2*x20, x2*x21, x11*x3, x12*x3, x13*x3, x14*x3, x15*x3, x16*x3, x17*x3, x18*x3, x19*x3, x20*x3, x21*x3, x11*x4, x12*x4, x13*x4, x14*x4, x15*x4, x16*x4, x17*x4, x18*x4, x19*x4, x20*x4, x21*x4, x11*x5, x12*x5, x13*x5, x14*x5, x15*x5, x16*x5, x17*x5, x18*x5, x19*x5, x20*x5, x21*x5, x11*x6, x12*x6, x13*x6, x14*x6, x15*x6, x16*x6, x17*x6, x18*x6, x19*x6, x20*x6, x21*x6, x11*x7, x12*x7, x13*x7, x14*x7, x15*x7, x16*x7, x17*x7, x18*x7, x19*x7, x20*x7, x21*x7, x11*x8, x12*x8, x13*x8, x14*x8, x15*x8, x16*x8, x17*x8, x18*x8, x19*x8, x20*x8, x21*x8, x11*x9, x12*x9, x13*x9, x14*x9, x15*x9, x16*x9, x17*x9, x18*x9, x19*x9, x20*x9, x21*x9, x10*x11, x10*x12, x10*x13, x10*x14, x10*x15, x10*x16, x10*x17, x10*x18, x10*x19, x10*x20, x10*x21, x0^2, x1^2, x2^

We save `X_mon2` to `X_mon2.sobj` for future use.

In [4]:
save(X_mon2, "X_mon2")

An abritrary symmetric matrix $H$ of size $143 \times 143$ has $\frac{143 \cdot 144}2 = 10296$ free parameters. 

In [5]:
H = matrix(SR, X_mon2_len, X_mon2_len)
H_vars = [var("h"+str(i)) for i in range(X_mon2_len * (X_mon2_len + 1) / 2)]
count = 0
for i in range(X_mon2_len):
    for j in range(i,X_mon2_len):
        H[i, j] = H[j, i] = H_vars[count]
        count += 1
print(len(H_vars))
print("")
print(H.is_symmetric())
print("")
H[130:,130:] # example

10296

True



[h10205 h10206 h10207 h10208 h10209 h10210 h10211 h10212 h10213 h10214 h10215 h10216 h10217]
[h10206 h10218 h10219 h10220 h10221 h10222 h10223 h10224 h10225 h10226 h10227 h10228 h10229]
[h10207 h10219 h10230 h10231 h10232 h10233 h10234 h10235 h10236 h10237 h10238 h10239 h10240]
[h10208 h10220 h10231 h10241 h10242 h10243 h10244 h10245 h10246 h10247 h10248 h10249 h10250]
[h10209 h10221 h10232 h10242 h10251 h10252 h10253 h10254 h10255 h10256 h10257 h10258 h10259]
[h10210 h10222 h10233 h10243 h10252 h10260 h10261 h10262 h10263 h10264 h10265 h10266 h10267]
[h10211 h10223 h10234 h10244 h10253 h10261 h10268 h10269 h10270 h10271 h10272 h10273 h10274]
[h10212 h10224 h10235 h10245 h10254 h10262 h10269 h10275 h10276 h10277 h10278 h10279 h10280]
[h10213 h10225 h10236 h10246 h10255 h10263 h10270 h10276 h10281 h10282 h10283 h10284 h10285]
[h10214 h10226 h10237 h10247 h10256 h10264 h10271 h10277 h10282 h10286 h10287 h10288 h10289]
[h10215 h10227 h10238 h10248 h10257 h10265 h10272 h10278 h10283 h10287 h10290 h10291 h10292]
[h10216 h10228 h10239 h10249 h10258 h10266 h10273 h10279 h10284 h10288 h10291 h10293 h10294]
[h10217 h10229 h10240 h10250 h10259 h10267 h10274 h10280 h10285 h10289 h10292 h10294 h10295]

Now we definie $G = \boldsymbol x^T H \boldsymbol x$ and the we solve the linear system $F - G = 0$. 

In [6]:
G = (X_mon2 * H * X_mon2).expand()

We start by defining the relevant monomials of degree $4$.

In [7]:
X_mon4 = list({mon1 * mon2 for mon1 in X_mon2 for mon2 in X_mon2})
print(len(X_mon4))

7150


In [8]:
%time F_dict = {mon: F.coefficient(mon) for mon in X_mon4}

CPU times: user 1.11 s, sys: 71 μs, total: 1.11 s
Wall time: 1.11 s


In [9]:
# long time
%time G_dict = {mon: G.coefficient(mon) for mon in X_mon4}

CPU times: user 28.1 s, sys: 6.43 ms, total: 28.1 s
Wall time: 28.1 s


In [10]:
%time H_eqs_dict = {mon: F_dict[mon] - G_dict[mon] == 0 for mon in X_mon4}
H_eqs = list(H_eqs_dict.values())

CPU times: user 57.9 ms, sys: 20 μs, total: 57.9 ms
Wall time: 57.4 ms


In [11]:
# long time
%time H_sols = solve(H_eqs, H_vars, algorithm="sympy")

CPU times: user 23.6 s, sys: 82.7 ms, total: 23.7 s
Wall time: 23.9 s


Now we save the general solution for future use. Making the substitution `H.subs(H_sols)` is very expensive but here's a workaround.

In [12]:
H_sols_dict = H_sols[0]
H_sols_keys_list = list(H_sols_dict.keys())

H1_vars = H_vars.copy()

H_vars_index = {}
count = 0
for hi in H_vars:
    H_vars_index[hi] = count
    count += 1
    
for k in H_sols_keys_list:
    ind = H_vars_index[k]
    H1_vars[ind] = H_sols_dict[k]

H1 = matrix(SR, X_mon2_len, X_mon2_len)
count = 0
for i in range(X_mon2_len):
    for j in range(i,X_mon2_len):
        H1[i, j] = H1[j, i] = H1_vars[count]
        count += 1
        
H1[30:40,30:40] # example

[    -2*h10103 + 3/4*r                     0                     0                 h3858                 h3859                 h3860                 h3861                 h3862                 h3863                 h3864]
[                    0     -2*h10104 + 3/4*r                     0                 h3970                 h3971                 h3972                 h3973                 h3974                 h3975                 h3976]
[                    0                     0     -2*h10105 + 3/4*r                 h4081                 h4082                 h4083                 h4084                 h4085                 h4086                 h4087]
[                h3858                 h3970                 h4081       -2*h10114 + 1/2                     0                     0                     0                     0                     0                     0]
[                h3859                 h3971                 h4082                     0       -2*h10115 + 1/2                     0                     0                     0                     0                     0]
[                h3860                 h3972                 h4083                     0                     0     -2*h10116 + 3/4*r                     0                     0                     0                     0]
[                h3861                 h3973                 h4084                     0                     0                     0             -2*h10117                     0                     0                     0]
[                h3862                 h3974                 h4085                     0                     0                     0                     0       -2*h10118 + 1/2                     0                     0]
[                h3863                 h3975                 h4086                     0                     0                     0                     0                     0       -2*h10119 + 1/2                     0]
[                h3864                 h3976                 h4087                     0                     0                     0                     0                     0                     0 -2*h10120 - 9/4*r + 2]

We check for symmetry.

In [13]:
H1.is_symmetric()

True

We check that `H1` is the desired solution.

In [14]:
(X_mon2 * H1 * X_mon2 - F).expand()

0

In [15]:
save(H1, "H_gen_matrix")